File: edukate-common/src/main/java/io/github/sanyavertolet/edukate/common/users/UserRole.java
package io.github.sanyavertolet.edukate.common.users;

import org.springframework.security.core.authority.SimpleGrantedAuthority;

import java.util.Arrays;
import java.util.Collection;
import java.util.EnumSet;
import java.util.Set;
import java.util.stream.Collectors;

public enum UserRole {
    USER,
    MODERATOR,
    ADMIN,
    ;

    private static final String ROLE_PREFIX = "ROLE_";

    public String asSpringSecurityRole() {
        return ROLE_PREFIX + name();
    }

    public SimpleGrantedAuthority asGrantedAuthority() {
        return new SimpleGrantedAuthority(asSpringSecurityRole());
    }

    public static Set<UserRole> fromString(String rolesString) {
        if (rolesString == null || rolesString.trim().isEmpty()) {
            return EnumSet.noneOf(UserRole.class);
        }
        return Arrays.stream(rolesString.split(","))
                .map(String::trim)
                .map(UserRole::valueOf)
                .collect(Collectors.toCollection(() -> EnumSet.noneOf(UserRole.class)));
    }

    public static String toString(Collection<UserRole> roles) {
        return roles.stream()
                .map(UserRole::name)
                .collect(Collectors.joining(","));
    }

    public static Collection<UserRole> anyRole() {
        return EnumSet.allOf(UserRole.class);
    }
}

File: edukate-common/src/main/kotlin/io/github/sanyavertolet/edukate/common/users/UserRole.kt
package io.github.sanyavertolet.edukate.common.users

import org.springframework.security.core.authority.SimpleGrantedAuthority
import java.util.EnumSet

enum class UserRole {
    USER,
    MODERATOR,
    ADMIN;

    fun asSpringSecurityRole(): String = ROLE_PREFIX + name

    fun asGrantedAuthority() = SimpleGrantedAuthority(asSpringSecurityRole())

    companion object {
        private const val ROLE_PREFIX = "ROLE_"

        @JvmStatic fun listToString(roles: Set<UserRole>): String = roles.joinToString(",") { it.name }

        @JvmStatic
        fun fromString(rolesString: String?): Set<UserRole> =
            rolesString?.split(",")?.map { UserRole.valueOf(it.trim()) }?.toSet() ?: EnumSet.noneOf(UserRole::class.java)

        @JvmStatic fun anyRole(): Set<UserRole> = EnumSet.allOf(UserRole::class.java)
    }
}
